lib/client/index.js

```diff
@@ -0,0 +1,1710 @@
+    // required to happen when 'now' event is sent from websocket.js every minute.  When fixed,
+    // remove this code and all references to `type: 'server-forecast'`
+    var last = _.last(data);
+    var lastTime = last && last.mills;
+    if (!lastTime) {
+      console.error('Bad Data, last point has no mills', last);
+      lastTime = Date.now();
+    }
+
+    var n = Math.ceil(12 * (1 / 2 + (now - lastTime) / SIXTY_MINS_IN_MS)) + 1;
+    for (var i = 1; i <= n; i++) {
+      data.push({
+        mills: lastTime + (i * FIVE_MINS_IN_MS), mgdl: 100, color: 'none', type: 'server-forecast'
+      });
+    }
+  }
+
+// clears the current user brush and resets to the current real time data
+  function updateBrushToNow(skipBrushing) {
+
+    // get current time range
+    var dataRange = d3.extent(data, dateFn);
+
+    // update brush and focus chart with recent data
+    d3.select('.brush')
+      .transition()
+      .duration(UPDATE_TRANS_MS)
+      .call(brush.extent([new Date(dataRange[1].getTime() - foucusRangeMS), dataRange[1]]));
+
+    addPlaceholderPoints();
+
+    if (!skipBrushing) {
+      brushed(true);
+
+      // clear user brush tracking
+      brushInProgress = false;
+    }
+  }
+
+  function brushStarted() {
+    // update the opacity of the context data points to brush extent
+    context.selectAll('circle')
+      .data(data)
+      .style('opacity', 1);
+  }
+
+  function brushEnded() {
+    // update the opacity of the context data points to brush extent
+    context.selectAll('circle')
+      .data(data)
+      .style('opacity', function (d) { return highlightBrushPoints(d) });
+  }
+
+  function alarmingNow() {
+    return container.hasClass('alarming');
+  }
+
+  function inRetroMode() {
+    if (!brush) {
+      return false;
+    }
+
+    var time = brush.extent()[1].getTime();
+
+    return !alarmingNow() && time - TWENTY_FIVE_MINS_IN_MS < now;
+  }
+
+  function brushed(skipTimer) {
+
+    if (!skipTimer) {
+      // set a timer to reset focus chart to real-time data
+      clearTimeout(brushTimer);
+      brushTimer = setTimeout(updateBrushToNow, BRUSH_TIMEOUT);
+      brushInProgress = true;
+    }
+
+    var brushExtent = brush.extent();
+
+    // ensure that brush extent is fixed at 3.5 hours
+    if (brushExtent[1].getTime() - brushExtent[0].getTime() !== foucusRangeMS) {
+      // ensure that brush updating is with the time range
+      if (brushExtent[0].getTime() + foucusRangeMS > d3.extent(data, dateFn)[1].getTime()) {
+        brushExtent[0] = new Date(brushExtent[1].getTime() - foucusRangeMS);
+        d3.select('.brush')
+          .call(brush.extent([brushExtent[0], brushExtent[1]]));
+      } else {
+        brushExtent[1] = new Date(brushExtent[0].getTime() + foucusRangeMS);
+        d3.select('.brush')
+          .call(brush.extent([brushExtent[0], brushExtent[1]]));
+      }
+    }
+
+    var nowDate = new Date(brushExtent[1] - THIRTY_MINS_IN_MS);
+
+    function updateCurrentSGV(entry) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l337)